applications/sentence-transformers-ft/modal-optuna.py

```diff
@@ -0,0 +1,135 @@
+from sentence_transformers import SentenceTransformer
+from datasets import load_dataset
+import modal
+import pathlib
+import optuna
+from finetune import finetune
+from datetime import datetime
+import torch
+
+
+MODEL_ID = "BAAI/bge-small-en-v1.5"
+
+# Modal constants
+GPU_CONFIG = "a10g"
+N_GPU = 4
+N_TRIALS = 3  # Number of trials PER gpu. Total trials = N_GPU * N_TRIALS
+USE_CACHED_IMAGE = True  # enable this to download the dataset and base model into the image for faster repeated runs
+
+
+# Functions for Modal Image build steps (cache the model and dataset)
+def download_model():
+    SentenceTransformer(MODEL_ID)
+
+
+def download_dataset():
+    dataset_id = "quora"
+    load_dataset(dataset_id, split="train")
+
+
+# Modal resources
+stub = modal.Stub("finetune-embeddings-optuna")
+image = modal.Image.debian_slim().pip_install(
+    "sentence-transformers", "torch", "datasets", "optuna"
+)
+if USE_CACHED_IMAGE:
+    image = image.run_function(download_model).run_function(download_dataset)
+
+
+# TODO: attach this to stub?
+nfs_volume = modal.NetworkFileSystem.new()
+JOURNAL_PATH = "/root/cache/journal.log"
+STUDY_NAME = "sentence-transformers-ft study"
+VOL_MOUNT_PATH = pathlib.Path("/vol")
+
+# TODO: uncomment this to make the volume persisted
+volume = modal.Volume.persisted(
+    f"sentence-transformers-ft-optuna-{int(datetime.now().timestamp())}"
+)
+# non-persistent volume for dev purposes
+# volume = modal.Volume.new()
+
+
+@stub.function(image=image, network_file_systems={"/root/cache": nfs_volume})
+def initialize_optuna():
+    storage = optuna.storages.JournalStorage(
+        optuna.storages.JournalFileStorage(JOURNAL_PATH)
+    )
+    storage.create_new_study(
+        study_name=STUDY_NAME, directions=[optuna.study.StudyDirection.MAXIMIZE]
+    )
+
+
+def objective(trial: optuna.Trial):
+    # TODO: set more interesting hyperparameter searches
+    dense_out_features = trial.suggest_int(
+        "dense_out_features", 200, 800, log=True
+    )  # TODO: Need to also somehow suggest "None" for running without dense layer, or this could just be manually set by user in the script
+    activation_function_str = trial.suggest_categorical(
+        "activation", ["Tanh", "ReLU", "Sigmoid"]
+    )
+    activation_function = getattr(torch.nn, activation_function_str)()
+    epochs = trial.suggest_int("epochs", 7, 12, log=True)
+
+    scheduler = trial.suggest_categorical(
+        "scheduler", [
+            "warmupconstant", "warmuplinear", "warmupcosine"
+        ] 
+    )
+    # TODO: add dropout, seems kinda annoying tho https://github.com/UKPLab/sentence-transformers/issues/677
+
+    # TODO: add learning rate scheduler
+    res = finetune(
```

Comment:
Might need to modify this to split dataset into "train" "validation" and "test". The model will train on "train" data, Optuna will do grid search using the score of the "validation" dataset, and the final evaluation will be done on "test".

https://stats.stackexchange.com/questions/366862/is-using-both-training-and-test-sets-for-hyperparameter-tuning-overfitting